### ISSUE TITLE
X9D+ SE 2019 and multimodule communication cut off with SF switch when openTX >= 2.3.12

### ISSUE THREAD

**Issue by EddyBag42** [opentx/opentx#8786]
Hi, 

I have a Taranis X9D+ SE 2019 (openTX 2.3.14) and an iRangeX IRX4+ multimodule (firmware 1.3.6)
I'm facing a fairly critical issue of communication cut off between radio and module.

In serial mode (MULTI), communication between the radio and the multimodule cuts off under reproducible circumstances as follows :
- radio is using external RF 
- communication between radio and module is set in serial mode (MULTI)
- receiver doesn't need to be on (though problem may appear quicker when receiver is on, but not sure of this point)
- after several toggles of the 2-positions switch SF (random number of toggles, that can be one or ten, rarely more than a few tens) the communication between the radio and the multimodule cuts off (the radio says "No serial input" and the red led flashes slowly on the multimodule)

![image](https://user-images.githubusercontent.com/9478418/143895294-a130a1f3-ba6e-49ca-9fd3-125d161386b2.png)

I posted this issue to RC Groups thread and it appears that this is not an isolated case, it seems to be related to using multimodule inside X9D+ SE 2019, please [see here](https://www.rcgroups.com/forums/showpost.php?p=48240461&postcount=23142).

[Old openTX versions](https://www.rcgroups.com/forums/showpost.php?p=48249615&postcount=23150) for X9D+ SE 2019 have been posted on the same thread so I could try and see if issue is still there.

Here are the results of my tests : 
- openTX 2.3.13 : issue is present
- openTX 2.3.12 : issue is present
- openTX 2.3.11 : no issue detected
- openTX 2.3.10 : no issue detected
- openTX <= 2.3.9 : not tested

So it seems to me that the issue appeared with openTX 2.3.12 and is there since.

Thanks for your help. Best regards.

Remark : no issue with another switch than SF, no issue with PPM communication mode, no issue with SF switch when using internal RF, no issue with an X9D, seems specific to X9D+ SE 2019 + multimodule + serial mode + SF switch

**Comment by EddyBag42** [comment]
Cut off bug still present with openTX 2.3.15

**Comment by gratefulfrog** [comment]
I confirm that this bug is there and terrible

**Comment by gratefulfrog** [comment]
any progress??

Thanks,
B

**Comment by gratefulfrog** [comment]
Hello?  Is there any way to help you guys with this issue? It is really hurting me...

**Comment by HippoDan** [comment]
This is still an ongoing problem.  Lost a plane this week because of it.  Going back to 2.3.11 for the time being.  Please fix this, I don't want to be stuck on older firmware forever.

**Comment by bsongis** [comment]
Do you use SF for anything? Like screenshot

**Comment by gratefulfrog** [comment]
Yes, I use SF for throttle cut on all my models.

Thanks for your help, it would be great to fix this so we can move to the
latest version of open-tx!

**Comment by bsongis** [comment]
Would you share your eeprom here?

**Comment by bsongis** [comment]
Do you reproduce if you use SF for nothing?
And on a blank model?

**Comment by gratefulfrog** [comment]
Yes, it still happens on a blank model with nothing assigned to SF.

**Comment by gratefulfrog** [comment]
I am happy to share my eeprom, but how do I do it? Is it available on the
SD card?

**Comment by HippoDan** [comment]
> Do you use SF for anything? Like screenshot

I like my momentary switch on the left, and I use it to call out my timer among other things.  It gets pressed a lot for timer, gun sound effects, self level and more.  Started having the problem right away, but it took forever for me to figure out what was happening.

**Comment by gratefulfrog** [comment]
So are you able to work on a bug fix? This problem has been open for a
while and no one seems to be looking at it...

I would look at the code, but it would take me so long to dive in... I
hardly have the time.

Let me know about the EEPROM if you could use it - it is version 219 I think

**Comment by Ceeb182** [comment]
The EEPROM is the complete copy of the radio's contents (models + radio configuration + OpenTx firmware).

To obtain the EEPROM and attach it here in a message: 
1. Install Companion 2.3 software on a PC
2. Put the radio in "Bootloader" mode (press the two horizontal trims towards the center while powering up the radio, the screen will turn on with "OpenTx Bootloader" in the title bar).
3. Open the Companion software on PC (if the software suggests updates, ignore them for the moment, i.e. answer "No" twice).
4. Connect the radio via USB to the PC. The radio displays "USB Connected".
5. In Companion software, go to menu: Read Write > Backup Radio To File
6. Choose an explicit file name (Example: "OTX2313-BugSF-ModelName.bin")
7. Companion displays a bar graph showing the progress of the EEPROM transfer to the file "OTX2313-BugSF-ModelName.bin".
8. Attach the file to this issue's message (by dragging and dropping "OTX2313-BugSF-ModelName.bin" in the message)
9. Specify in the message: radio type, radio's OpenTx version, model name.
10. And that's it!

**Comment by bsongis** [comment]
Someone reproduces the issue with a radio with the small external module bay? because I only have the small Multimodule ...

**Comment by gratefulfrog** [comment]
As requested, I attach my EEPROM File, but please note that this is OTX
version 2.3.11 which is the most recent version that DOES NOT HAVE THE BUG.

The bug is present in all versions 2.3.XX where XX > 11.

I do not use the more recent version because of the bug!

My firmware is:
FW: opentx-x9d+2019
VERS: 2.3.11-OTX (4328055e)
Date: 2021-01-08 23:44:58
EEPR: 219

I hope this helps, I am so glad that you guys are working on it. Let me
know if I can help, I am a good programmer and debugger...
Cheers,
Bob

**Comment by bsongis** [comment]
Searching for a module to reproduce the problem ...

**Comment by bsongis** [comment]
Now I have X9D+ updated to latest 'master', compiled with a recent compiler (the one I use for Ethos)
I have a multimodulle in version 1.3.3.20 AETR
I cannot reproduce the problem

**Comment by bsongis** [comment]
Now the same opentx, compiled with the old good compiler. I cannot reproduce the problem.

**Comment by bsongis** [comment]
Now the official firmware downloaded from companion. I cannot reproduce the problem.
I have not bound the RX, should I?
Which OpenTX options did you choose?

**Comment by bsongis** [comment]
@gratefulfrog can you also share your eeprom here?

**Comment by bsongis** [comment]
Now with the official firmware downloaded from companion and bound with a Spektrum AR410.
I cannot reproduce the problem

**Comment by EddyBag42** [comment]
On Fri, Sep 22, 2023, 06:44 Bertrand Songis ***@***.***>
wrote:

> Now I have X9D+ updated to latest 'master'
>
Hi,
Is it a X9D+ SE 2019 ?
You will not have thé problem with a regular X9D+.
Cheers.
-- 
Ed

**Comment by bsongis** [comment]
Yes of course

**Comment by bsongis** [comment]
![image](https://github.com/opentx/opentx/assets/5603504/2777e7ad-7f02-4394-880a-f37ddd0fe998)

**Comment by bsongis** [comment]
No luck, I have the multimodule marked as "No problem"
I also have the iRX4 lite, searching for a way to connect it to the fullsize module slot...

**Comment by 3djc** [comment]
> On Fri, Sep 22, 2023, 06:44 Bertrand Songis ***@***.***> wrote: Now I have X9D+ updated to latest 'master'
> Hi, Is it a X9D+ SE 2019 ? You will not have thé problem with a regular X9D+. Cheers. -- Ed

At this point in time, we cannot reproduce and we spend time going through all the differences between 2.3.11 and 2.3.12 and found nothing that could explain that

**Comment by 3djc** [comment]
> On Fri, Sep 22, 2023, 06:44 Bertrand Songis ***@***.***> wrote: Now I have X9D+ updated to latest 'master'
> Hi, Is it a X9D+ SE 2019 ? You will not have thé problem with a regular X9D+. Cheers. -- Ed

If you have the issue too, an eeprom back would be helpful

**Comment by bsongis** [comment]
Now testing with iRX4, bound to the AR410
Still no issue

**Comment by 3djc** [comment]
At this point in time, we cannot find any issue. Unless we get a non working eeprom, we will stop investigating

**Comment by gratefulfrog** [comment]
Hello,
I am sorry that you have not been able to reproduce the bug.

It simply requires a Taranis X9D+ 2019 SE and any version of OTX 2.3.12 or
more recent and an IRX4 multimodule with latest firmware.

I  could try to flash the new OTX but I am very worried that my models will
be ruined or lost when I do that.

Is there a procedure that would ensure that I could go back to the exact
same configuration that I now have? And are you sure that it will work?

I would be destroyed if I had to reprogram my radio and all my models...

Thanks for your help:
Cheers,
Bob

**Comment by EddyBag42** [comment]
gretefulfrog is perfectly right

I can't figure out why you can't reproduce this issue, it's incredibly
frustrating and too bad

I'll try to upgrade to latest firmware and see whether the issue is still
present or not. If I get the issue (which I think I will for now) I'll post
the eeprom as requested. I'll do that as soon as I can, in the next hours
or so.

Thanks

**Comment by EddyBag42** [comment]
As I've already said earlier in this thread, I confirm that SF bug is still there with openTX 2.3.15 and this particular radio and module.

[Here you have a zip](https://github.com/opentx/opentx/files/12703513/OTX2315-BugSF.zip) including 3 files :
- a picture of my radio with problem shown ("No serial input") so you can make sure the radio you're using is the same PLUS SE 2019 (carbon or blue, hall effect sticks and ACCESS)
- a settings and models backup made with "Backup radio to file"
- a firmware backup made with "Read firmware from radio"

You can use Model 15 named "test sf" for testing.

Hardware : X9D PLUS SE 2019 and iRangeX IRX4+ multimodule
Software : openTX 2.3.15

Thanks a lot for your investigations.
Hope it will help.

**Comment by bigtwin** [comment]
Maybe this is not related to the OTX-Version. May it is something different? An issue with hard- or firmware (none OTX)? Something which just and only strikes in >2.3.12 not because of an error in OTX, but some other option/feature/timing ...

**Comment by gratefulfrog** [comment]
What could explain that the problem only occurs from otx 2.3.12 onwards
other than a change in otx? There is nothing else that changed, is there? I
cannot see any other explanation.

In the 2.3.12 release notes I see a few points which are suspicious:
- add SI support to X9D 2019
- Always prioritize protocol data coming directly from the module
- Add support for module config (to allow two multimodule to use the same
ID for example)
- Update data tables to 1.3.2.84 (thanks MRC3742)

Perhaps there is some strange conflict created by these and/or other
changes?

**Comment by 3djc** [comment]
If it was a clear cut error, it would affect all multi and would be easy to reproduce. It is neither

**Comment by bigtwin** [comment]
This issue seems to be extremly specific ...

Switch SF ... may it has nothing to do with SF but the 6th switch or the 4th being used, or the one before SG, or the switch with a cabling that is close to some point at the main board. Or switch chattering is not prevented by an capasitor.

Exact this type of Transmitter? May all your guys trasmitters were manufactured the same time? Equipped with the same cabling/switch or assembled by the same person (critical cable placement)?

Since it is that specific, I would think this is not direct software related, maybe indirect. The code has changed since 2.3.12 and now some other critical condition strikes in and causes this malfunction. It is obviously sophisticated and "since 2.3.12" may leads to a wrong approach. If there is an interest to debug such a tricky malfunction, the troubleshooting certainly need to stay open minded.
I would even consider that this problem is not resolvable with available ressources and it may should be covered by documentation/PreReqs.

**Comment by EddyBag42** [comment]
It definitely has something to do with SF. Problem does not appear with other switches. I've done enough testing, and gratefulfrog, and the guy from RCGroups. It is a fact.

Another fact is that problem appeared with openTX 2.3.12, and is there since.

It is related to specific hardware that's right. But it's really hard for me to believe that only several Plus SE 2019 are affected !

Anyway a change in openTX revealed this issue. Maybe if we knew what change in openTX code revealed the issue this would help to understand. What other way is there ?

This issue is critical. Your risk loosing your model if you touch SF, even the first time you touch it ! It's not only a question of
documentation.

**Comment by bigtwin** [comment]
Yes, you are right. But having in mind what limited ressources are available for development and troubleshooting in OTX, it may is just impossible to identify the root cause. Yes, with unlimited ressources it would certainly be possible to resolve this problem technically. But those are not available, therefor documentation is surley a potential solution. May even the only one we can get.

You cannot blame the loss of a model to this bug. Especially not, since it is now a knwon issue. Have in mind that you are using a transmitter and a certain configuration (SF) eventhough you know about this bug. You may should simply stay away from SF or that kind of transmitter.

Dont get me wrong. I agree in all what you wrote, but as said - if it is not possible to identify the root cause, it is probably time to live with it and work around it. This is a common way you cannot ignore. Even commercial manufactures choose this sometimes. Take the _Profi TX_ mishap from _Multiplex_ as an example.  
I lost once a model due to some bug in _Spectrum_. Even 100% reproducable. It was committed and confirmed by the dealer and Spectrum, but never clarified to the end. I simply had to stay away from some types of receivers eventhough they were officially documented as being OK.

**Comment by 3djc** [comment]
We have reviewed with 2 pair of eyes EVERY single line of code different between 2.3.11 and 2.3.12, and found nothing that can explain that.

We also haven't been able to reproduce so far (yes, using very same radio and module). Another attempt will be done Monday, but no luck so far.

**Comment by EddyBag42** [comment]
Ok, too bad you couldn't reproduce... Maybe specific to some batch... Thanks a lot for your efforts anyway !

@bigtwin : sure it's openTX team's choice to stop or continue investigations

**Comment by 3djc** [comment]
We haven't give hope yet, but we already spent a lot of time on this issue without been able to find anything

**Comment by gratefulfrog** [comment]
I appreciate all that has been said and understand that resources are
limited.

What could explain the problem, if it's not the code? Some bad radios
coincidentally purchased? I myself an a professional in IT and have leadsny
projects. My experience is that the problem is always the code.

Is there anything I could do to help?

Did the compiler options change from 2.3.11 to 2.3.12 by any chance?

**Comment by 3djc** [comment]
I don't understand that comment on ressource, it doesn't apply here at all.

We reviewed code changes line by line: nothing suspect.

We try to reproduce: impossible so far.

So at this point in time, until we find a way to reproduce (if possible at all), we have no way forward

**Comment by gratefulfrog** [comment]
I'm sorry, I meant no offense.

And you cannot reproduce the problem with an frsky x9d+ 2019SE?

I don't know how to proceed either, here we reproduce the problem 100% of
the time.

What can we do to help?

**Comment by kilrah** [comment]
> Hardware : X9D PLUS SE 2019 and iRangeX IRX4+ multimodule Software : openTX 2.3.15

What multi firmware? Has it been confirmed across multiple multi firmware versions?
Do the multis exhibiting the problem have something in common e.g. the smaller and/or clone MCU?

**Comment by bsongis** [comment]
My multimodule is the iRX4 Lite, flashed with Multi V1.3.2.61AETR
I have flashed the firmware from your ZIP
I have transfered the eeprom from your ZIP
I have bound my RX as you also use DSM
I have toggled SF around 100 times, at various speeds.
Everything works perfectly on my radio, the link is rock solid

**Comment by gratefulfrog** [comment]
Hi,

Thanks for allthis work!

Which zip EEPROM did you use exactly? Mine (gratefulfrog) is working since
it is otx 2.3.11, the zip from EddyBag42 is the one with the bug.

I cannot understand how it could work when I did it, after 3 or 4 cliccks
on SF, the comm is lost?

**Comment by 3djc** [comment]
We tried over 100 SF toggle, all works perfect, despite using the supplied firmware, eeprom, everything

**Comment by bsongis** [comment]
@gratefulfrog only EddyBag42 attached an EEPROM here. I installed it with the firmware provided in the same ZIP.
I cannot reproduce the problem. Now I doubt it's only a firmware issue.

**Comment by bsongis** [comment]
I have pushed the problem to FrSky hardware designer, I hope he will have an idea

**Comment by gratefulfrog** [comment]
I really appreciate all your efforts!

I just can't find an explanation of how it could fail for both me and Eddy
and a third guy, yet for no one else...

   - our radios are all defective?  would seem unlikely
   - our multimodules defective, again unlikely,
   - maybe your radio is a more recent version of the FrSky X9D+ 2019 SE
   and the hardware is different? But the 2019 SE is only a few years old,
   could FrSky have already changed the hardware spec?

As I said, my experience with decades of software development is that the
problem is always in the code, but this time I am at a loss as to how to
explain that you guys don't have the issue...

I wish I could help....

Cheers,
Bob

**Comment by HippoDan** [comment]
Mine is also an IRX4 module on 1.3.3.20.  I can borrow a friend's Jumper module on Wednesday and see if it has the same problem.  I'll upload an Eeprom then.

**Comment by gratefulfrog** [comment]
I'm wondering if this whole problem could be a result of the SF switch
bouncing more or less on certain radios? Maybe FrSky changed switches, or
modified the filtering circuit at some point and the bounce is causing
problems for the multi-module ?

Just a thought, based on decades of hardware hacking ;-)

Cheers,
B

**Comment by HippoDan** [comment]
I'm not using the stock switch, I replaced it with a momentary pushbutton.  

> I'm wondering if this whole problem could be a result of the SF switch bouncing more or less on certain radios? Maybe FrSky changed switches, or modified the filtering circuit at some point and the bounce is causing problems for the multi-module ? Just a thought, based on decades of hardware hacking ;-) Cheers, B
>

**Comment by gratefulfrog** [comment]
Wouldn't that possibly explain why the bug can't seem to be reproduced?

**Comment by HippoDan** [comment]
Possibly.  I saw it as more of a reason to eliminate the switch as mine is a different make and it still happens.  It still doesn't explain why it happens only in software 2.3.12 and up.

**Comment by HippoDan** [comment]
I'll try to test my friend's Jumper 4-in-1 module with my 2.3.15 Taranis 2019 SE today.   Also, has anyone done this mod to their module?  I'm considering trying that next.

https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/issues/621

**Comment by HippoDan** [comment]
I borrowed the Jumper module yesterday.  I clicked over 100 times without the problem.  Turn off/on the radio, clicked 120 times, no problem.  I put the IRX4 back in and clicked 100 times no problem, off/on, then 120 times no problem.  Very confused.  Repeated and the problem reappeared at 50 clicks, then 21 clicks.

Now I don't know whether the Jumper module was a lucky break, or not having the problem.  Intermittent problems are the worst.  How will I ever know if I can trust the radio again?

Anyway, when I see him again I'll test the Jumper module again.  Anybody else have access to a different module?

**Comment by EddyBag42** [comment]
Jumper module has been tested with no problem, see [here](https://www.rcgroups.com/forums/showpost.php?p=48240461&postcount=23142).

**Comment by HippoDan** [comment]
Strangely it seems the protocol has an effect.  On FrSky V mode it takes anywhere from 1 to 25 presses to cause the problem with it rarely going over that.  On the v761 protocol it usually takes 25-60 presses, with rare outliers for the problem to happen.  I guess the complexity/quantity of data is affecting the outcome.  Some kind of buffer overflow?  Hardware is not my area.

**Comment by 3djc** [comment]
The dialog between mpm module and radio does not change with protocol, and also is exactly the same for all brands of mpm

**Comment by gratefulfrog** [comment]
Hello Guys,

Today I decided to take a risk and update my Frskly Taranis X9D plus 2019
SE to the latest firmware and see if I could reproduce the bug.

The short version is :

   - Yes, the Bug is present : after 46 clicks of the SF switch, the module
   status becomes "no serial input" and the red led starts flashing.
   - Yes, I confirmed this by restarting the radio and again after 59
   clicks, "no serial input".
   - I attach the EEPROM file which shows the bug in the model MultTest
   with the IRangeX Multimodule

Here is the exact process that I followed:

   1. using companion, I downloaded the latest version of  OTX: 2.3.15 and
   configured the options : opentx-x9d+2019-afhds3-flexr9m-lua-en-2.3.15-otx.
   2. I downloaded the content of the corresponding SD card
   3. I  made a backup copy of my current SD card
   4. I made a backup of my models and settings
   5. I made a backup of my radio
   6. I copied the content of the new SD card to the physical card
   7. I added the files that were specific to my old sd card to the new
   one, putting them in the new directories as needed and deleting all the
   complied versions of the lua scripts
   8. I flashed the latest firmware to the radio with the companion.
   9. I restarted the radio, physically installed the IRangeX module and
   selected the MultTest model
   10.  I observed on the model configuration menu

External RF:
Mode : MULTI
Type Frsky D D8
Module Status  V1.3.3.0 RETA  <-- edited after original post (sorry)
Channel Range CH1-16
Receiver 20 [Bind] [Range]

   1. I observed that the red and green leds were lit and solid.
   2. I began to slowly click the SF switch.
   3. After 46 clicks the display showed:

Module Status  No serial input
the RED LED began to flash

   1. I shut down the radio
   2. I restarted the radio
   3. I observed the correct Module Status
   4. I began clicking the SF switch slowly
   5. After 59 clicks, the status changed to "no serial input" and the RED
   LED began flashing

I  believe that we can be sure that the bug is still present.

I sincerely hope that this helps with the search for the problem. Let me
know if there is anything I can do to help!

Thanks for all the effort that everyone is putting in!!
Cheers,
Bob

**Comment by 3djc** [comment]
Why do you have afhds3 ? Do you have such module ?

**Comment by gratefulfrog** [comment]
I do not even know what it is. It was checked in the companion, maybe
by default? or by accident?

**Comment by 3djc** [comment]
It is not defaulted, it has to be manually selected. Try without

**Comment by gratefulfrog** [comment]
I'm sure it will change nothing.
I already restored my radio... Do you really want me to do it?

**Comment by gratefulfrog** [comment]
ps, with version 2.3.11 restored, 120 clicks of SF and no bug produced...

**Comment by gratefulfrog** [comment]
Hello Team,

I've been looking at release diffs and had an idea.

Let me tell what I've been looking at :

   - First, I cloned the opentx repo from github to my pc
   - then I did some diffs:
      - $ cd opentx
      - get the names of the files and stats of the diff between release
      2.3.11 and 2.3.12, and save it to a file: *d_11_12*
      - $ git diff release/2.3.11  release/2.3.12 --stat > ../d_11_12
      - then do the same for 12 to 13, 13 to 14 and 14 to 15:
      - $ git diff release/2.3.12  release/2.3.13 --stat > ../d_12_13
      - $ git diff release/2.3.13  release/2.3.14 --stat > ../d_13_14
      - $ git diff release/2.3.14  release/2.3.15 --stat > ../d_14_15
   - Now looking at all the file names, let's select only those in the
   'radio/' directory, save that in a file called *radio_d*
   - cd ..
   - grep 'radio/'  d_*  > radio_d
   - and finally, let's look at all the lines with the word 'multi' and/or
   'module' and put that in a file called
*radio_multi_module_d (attached) *
   - grep 'module\|multi'  radio_d > radio_multi_module_d
   - let's look at it:
   - $ cat  radio_multi_module_d
   d_11_12: radio/src/io/multi_firmware_update.cpp             |   10 +-
   d_11_12: radio/src/pulses/modules_constants.h               |   23 +-
   d_11_12: radio/src/pulses/modules_helpers.h                 |   10 +-
   d_11_12: radio/src/pulses/multi.cpp                         |   81 +-
   d_11_12: radio/src/pulses/multi.h                           |    1 +
   d_11_12: radio/src/targets/horus/extmodule_driver.cpp       |   59 +-
   d_11_12: radio/src/targets/taranis/extmodule_driver.cpp     |   71 +-
   d_11_12: radio/src/telemetry/multi.cpp                      |  174 +-
   d_11_12: radio/src/telemetry/multi.h                        |   22 -
   d_12_13: radio/src/targets/horus/extmodule_driver.cpp       |    2 +-
   d_13_14: radio/src/targets/horus/extmodule_driver.cpp       | 16 +++--
   d_14_15: radio/sdcard/tango/SCRIPTS/WIZARD/multi.lua        |  436 ++
   d_14_15: radio/src/io/multi_firmware_update.cpp             |   18 +-
   d_14_15: radio/src/pulses/modules_constants.h               |    4 +-
   d_14_15: radio/src/pulses/modules_helpers.h                 |    2 +-
   d_14_15: radio/src/targets/tbs/extmodule_driver.cpp         |  416 ++
   d_14_15: radio/src/telemetry/multi.cpp                      |   44 +
   - We see that around 450 lines changed from 2.3.11 to 2.3.12 in files
   with names containing 'mutli' and/or 'module'
   - Now what if we create a branch from release 2.3.15 and undo those
   changes and test that build?
   - if it passes with the SF test, then progressively add in the changes
      until the culprit is identified.
      - if it fails the SF test, then who knows what to think, sigh.

I have not yet looked at the files in question and I fear that I would need
support to understand the code, and also to build the test firmware and
install it on my radio, but if anyone is willing to provide some support, I
am happy to continue searching for this bug. (I am (or was) an expert C++
programmer)

Thanks for your help!
Cheers,
Bob

**Comment by gagarinlg** [comment]
Making a backup and try if EdgeTX has fixed this in the latest release is probably a lot quicker.

**Comment by gratefulfrog** [comment]
Hi,
Thanks for the tip. Are you sure that I will be able to restore the OpenTX
backup, once I load EdgeTX?

But even if Edge TX works, wouldn't it be of interest to the OpenTX
community to get it to work as well? There is clearly a bug...

Cheers,
B

**Comment by gagarinlg** [comment]
Yes, you can go back.

Regarding the future and continued support of OpenTX other people should answer.

**Comment by jtaylor2** [comment]
If you know that 2.3.11 is good and 2.3.12 is bad and can build the firmware "git bisect" will find the exact commit that caused the problem.  Then you can look to see what changed in that commit.

**Comment by gratefulfrog** [comment]
Hi,
That sounds great , but sorry for being a git newbie, how do i that? I have
never built  the firmware . Git bisect sounds like a way forward, can you
elaborate, please?

**Comment by jtaylor2** [comment]
Git bisect does a binary search for the bad commit given a good and bad starting point.  It checks out  commit half way between the good and bad.  You compile and test it and tell git bisect if it is good or bad.  It then checks out one half way again and again you compile and test it.  This continues until you find the exact commit that caused the problem  On first run bisect tells you how many iterations it may take to isolate the problem.

The down side is that you must to be able to compile the firmware and test it to find the commit that caused the problem and it may take 8 or 10 iterations which would be time consuming.

The other problem is that there must be thousands of those transmitters and only a couple of you are having the problem and evidently no one that can compile the firmware has the problem.  This seems to indicate that the problem must be some kind of race condition between the firmware and some specific piece of hardware of the (very few) failing units, or possibly hardware modifications of the failing units (I think at least some of them have had switches moved or other changes) and if that is the case just changing the build environment may change the problem.

**Comment by gratefulfrog** [comment]
Well, it may not be so hard because the problem only occurs with a specific
model transmitter (frsky Taranis x9d plus 2019 se) combined with a specific
multimodule (IrangeX) and only after actionning the SF switch around 50
times. I'm guessing that combination is not too frequently occurring.

The guy who changed the sf switch still had the problem.

I'm still hopeful.

Maybe I can learn how to build and install the firmware...

**Comment by jtaylor2** [comment]
Good point about that combination not too frequently occurring, but neither bsongis nor 3djc could duplicate the problem.

Installing the firmware is no problem, that's what you do when you upgrade.  Setting up a build environment is another story.  Seeing if the problem exists in edgeTX would be interesting.

**Comment by gratefulfrog** [comment]
Well, I am now trying to build opentx 2.3 by following the only
instructions that seem to applicable to my Linux envt.
https://github.com/opentx/opentx/wiki/OpenTX-2.3-Docker-Build-Instructions

Sadly, there does not seem to be a Dockerfile or a dockerfile in the repo
and thus I get:

opentx/radio/util$ docker build -t opentx .
[+] Building 0.0s (2/2) FINISHED

 docker:default
 => [internal] load build definition from Dockerfile

     0.0s
 => => transferring dockerfile: 2B

     0.0s
 => [internal] load .dockerignore

    0.0s
 => => transferring context: 2B

    0.0s
ERROR: failed to solve: failed to read dockerfile: open
/var/lib/docker/tmp/buildkit-mount4026469731/Dockerfile: no such file or
directory

I am at a loss, for the moment, as to how to progress...

Any thoughts?

**Comment by gratefulfrog** [comment]
Ok, so I am guessing that the docker bulld envt is obsolete.

This page gave me some insight :
https://www.rcgroups.com/forums/showthread.php?3824801-Blog-1-Step-by-step-guide-to-building-%28custom%29-OpenTX-firmware-under-Ubuntu-Linux

I have done some hacking around:

   - installed the Arm Gnu Toolchain:
      - followed instructions here:
      https://lindevs.com/install-arm-gnu-toolchain-on-ubuntu
   - installed dfu-util cmake :
      - $ sudo apt install dfu-util cmake
      - cloned the opentx 2.3 branch:
      - $ mkdir TEMP_ARM
      - $ git clone --recursive -b 2.3 https://github.com/opentx/opentx.git
      - $ cd opentx
      - $ mkdir build_00
      - $ cd build_00
      - tried to execute cmake with some options from
   https://github.com/opentx/opentx/wiki/OpenTX-2.2-Linux-Build-Instructions
      - $ cmake -DPCB=X9D+ -DGVARS=YES -DLUA=YES -DDEBUG=YES
      -DCMAKE_BUILD_TYPE=Debug ../
      - but got lots of errors about missing qt files
   - installed what I thought were the missing qtfiles including:
   qtmultimedia5-dev  and lots of others
   - tried the cmake again
      - $ cmake -DPCB=X9D+ -DGVARS=YES -DLUA=YES -DDEBUG=YES
      -DCMAKE_BUILD_TYPE=Debug ../
      - seemed to work! no errors
   - checked the results:
      - $ cmake -LH ../ > dev-cmake-options.txt
      - $ cat dev-cmake-options.txt
      $ more dev-cmake-options.txt
      -- OpenTX 2.3.15 @656fec93
      -- Python found, version: 3.10.12
      -- Qt Version: 5.15.3
      -- Found FOX: /usr/include/fox-1.6
      -- Foxlib found at /usr/lib/x86_64-linux-gnu/libFOX-1.6.so
      -- Found SDL: /usr/lib/x86_64-linux-gnu/libSDL.so (found version
      "1.2.15")
      -- SDL Lib: /usr/lib/x86_64-linux-gnu/libSDL.so; Headers:
      /usr/include/SDL; Version: 1.2.15
      -- Googletest will be downloaded
      -- Added optional gtests target
      -- Simulators library search path: /usr/local/lib/companion23
      -- Added optional gtests-companion target
      -- install /home/bob/TEMP_ARM/opentx/build_00 to /usr/local/bin
      -- Configuring done
      -- Generating done
      -- Build files have been written to:
      /home/bob/TEMP_ARM/opentx/build_00
      -- Cache values
      [etc]
   - tried to make it after looking at the targets in the Makefile with
   'make help'
      - $ make all
      - no errrors, but no firmware
      - $ make firmware
      - no errors:
   - this is the content of the directory after the make all and make
   firmware
   total 201588
   drwxrwxr-x  5 bob bob      4096 oct.  19 12:22 .
   drwxrwxr-x 13 bob bob      4096 oct.  19 11:34 ..
   -rw-rw-r--  1 bob bob     29542 oct.  19 12:08 CMakeCache.txt
   drwxrwxr-x  8 bob bob      4096 oct.  19 12:21 CMakeFiles
   -rw-rw-r--  1 bob bob      1891 oct.  19 11:45 cmake_install.cmake
   drwxrwxr-x  3 bob bob      4096 oct.  19 11:38 companion
   -rwxrwxr-x  1 bob bob 122792592 oct.  19 12:20 companion23
   -rw-r--r--  1 bob bob      3773 oct.  19 11:45 CPackConfig.cmake
   -rw-r--r--  1 bob bob      4193 oct.  19 11:45 CPackSourceConfig.cmake
   -rw-rw-r--  1 bob bob     10046 oct.  19 12:08 dev-cmake-options.txt
   -rw-rw-r--  1 bob bob     32768 oct.  19 12:21 eeprom.bin
   -rwxrwxr-x  1 bob bob    453372 oct.  19 12:21 firmware.bin
   -rwxrwxr-x  1 bob bob   6241424 oct.  19 12:21 firmware.elf
   -rwxrwxr-x  1 bob bob   8492960 oct.  19 12:13
   libopentx-x9d+-simulator.so
   -rw-rw-r--  1 bob bob     48963 oct.  19 12:09 Makefile
   drwxrwxr-x  3 bob bob      4096 oct.  19 11:35 radio
   -rwxrwxr-x  1 bob bob   6944504 oct.  19 12:13 simu
   -rwxrwxr-x  1 bob bob  61321008 oct.  19 12:20 simulator23
   -rw-rw-r--  1 bob bob       307 oct.  19 11:38 version.h
   - And this is the content of version.h:
   $ cat version.h
   #define VERSION           "2.3.15"
   #define VERSION_MAJOR     2
   #define VERSION_MINOR     3
   #define VERSION_REVISION  15
   #define VERSION_SUFFIX    ""

   #define VERSION_NUMBER    ((VERSION_MAJOR << 16) | (VERSION_MINOR << 8)
   | VERSION_REVISION)

   #define SIMULATOR_LIB_SEARCH_PATH  "/usr/local/lib/companion23"

So, now I am capable of building firmware.

Now, what I need:

   - *Which options* to choose to get the correct firmware for my Frsky
   Taranis X9D Plus 2019 SE, with LBT, i.e.  fix this line:
      - cmake -DPCB=X9D+ -DGVARS=YES -DLUA=YES -DDEBUG=YES
      -DCMAKE_BUILD_TYPE=Debug ../
   - How to use git to *select the propper files* to get the version 2.3.11
   then 2.3.12 ?
   - *How to use git bisect *exactly?
   - How to use* the output of git bisect to select new file sets* for
   building?

This is all a result of Jim Taylor's suggestion, Thank you Jim!

If possible, I would really appreciate a little help with the cmake options
and the git commands, if anyone has a minute?

Cheers,
Bob

**Comment by jtaylor2** [comment]
What version of Linux are you using?  I'm use openSUSE 15.4,
I would first compile 2.3.12 and make sure it still fails .
git tag will give you the tags so "git checkout release/2.3.12" should get you 2.3.12.

Looking at https://github.com/opentx/opentx/blob/2.3/tools/boards.py it looks like    you need -DPCB=X9D+ -DPCBREV=2019 

I would not do a debug build at this point, I would do a release build to match what you are downloading  and verify it fails when built with your environment 
-DCMAKE_BUILD_TYPE=RELEASE.

While you can do a make all you only need to do make -JX firmware (where X is the number of  computer cores to use to compile - the more you use the faster the compile).  

"git help bisect" will tell you how to use it.

**Comment by gratefulfrog** [comment]
I am including Juan G. from Frsky in this discussion since he has been
interested in this work.

Welcome Juan!

Have a look at the thread!
Cheers,
Bob

**Comment by HippoDan** [comment]
I will struggle to get a build environment going on a mac, and then go commit by commit until it breaks.  I don't know exactly when I'll have time, but I'm not giving up.

**Comment by gratefulfrog** [comment]
I tired that, using git bisect, but the intermediate firmware builds all
were bad, and wouldn't allow the radio to start...

However, when I rebuilt 2.3.15 and tried it, the bug was gone!!  I then
tried the companion build of the 2.3.15 and the bug was still gone.

Sadly, my friend EddyBag42 went through the same steps, all failed, but
that did not fix the bug for him...

I don't understand how this could be, and at a loss as to how I can further
debug, since I can no longer reproduce the bug...
